feat(ngcc): pause async processing if another process has the lockfile

petebacondarwin opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin committed:
feat(ngcc): pause async ngcc processing if another process has the lockfile

ngcc uses a lockfile to prevent two ngcc instances from executing at the
same time. Previously, if a lockfile was found the current process would
error and exit.

Now, when in async mode, the current process is able to wait for the previous
process to release the lockfile before continuing itself.

gkalpak commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -119,3 +86,105 @@ export class LockFile {
+ * when being called in an asynchronous context.
+ *
+ * * When ngcc starts executing, it creates a file in the `compiler-cli/ngcc` folder.
+ * * If it finds one is already there then it fails with a suitable error message.
```
No, it doesn't :wink:

gkalpak commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -119,3 +86,105 @@ export class LockFile {
+  }
+
+  protected async create() {
+    for (let attempts = 0; attempts < this.retryAttempts; attempts++) {
```
I think we should track `pid` and reset `attempts` if the `pid` changes. If there more than 2 ngcc processes waiting for each other, one of them will have to wait for all others to complete.

gkalpak reviewed (changes requested):
LGTM, except I think we should reset the waiting period when the `pid` in the lockfile changes.
Really excited to incorporate this into the CLI workflow \o/

petebacondarwin committed:
fixup! feat(ngcc): pause async ngcc processing if another process has the lockfile

petebacondarwin committed:
fixup! feat(ngcc): pause async ngcc processing if another process has the lockfile

petebacondarwin committed:
refactor(compiler-cli): add `invalidateCaches` to `CachedFileSystem`

This is needed by ngcc when reading volatile files that may
be changed by an external process (e.g. the lockfile).

gkalpak commented on packages/compiler-cli/ngcc/test/execution/lock_file_spec.ts:
```diff
@@ -335,10 +335,40 @@ runInEachFileSystem(() => {
+        ]]);
+        // We need to write to the rawFs to ensure that we don't update the cache at this point
+        rawFs.writeFile(lockFile.lockFilePath, '444');
+        await new Promise(resolve => setTimeout(resolve, 250));
```
This looks like a potential source of flakes :grin:
Could we use [fakeAsync()][1] and [tick()][2] to make this test deterministic?

[1]: https://angular.io/api/core/testing/fakeAsync
[2]: https://angular.io/api/core/testing/tick

petebacondarwin replied:
Yeah, I was a little bit worried about this. We cannot use the Angular `fakeAsync` stuff, since that is  driven via Zone.js and we definitely not going to bring that in here just for testing...

I tried to make the timeouts as reliable as possible so that it would be extremely unlikely for it to flake:

* having proven that the class has already checked the lockfile (and read its value) we are immediately writing the new one synchronously so there is no chance for the `setTimeout` inside `LockFile` to be triggered.
* we then set a timeout for 250ms before the next check but the class has been configured to check the contents of the lockfile every 100ms. So there should be at least 2 opportunities for a `setTimeout` to be triggered before this testing one is triggered.
* The class is setup not to fail until 10 attempts have occurred, which is a full second, so there is very little chance that the 250ms timeout is not triggered before we run out of attempts.

Do you have any better ideas?

gkalpak replied:
Fair enough. I am fine leaving this as is and address it in the future if we start seeing flakes.

For future reference, one way to make this deterministic would be using Jasmine's [Clock][1]. There is one twist, because we need to also flush microtasks after each `retryDelay` ms, so it would be something like:

```ts
const clock = jasmine.clock();
const moveToNextIteration = () => {
  clock().tick(100);     // Or whatever `retryDelay` we have specified.
  return Promise.resolve();  // To wait for microtask queue to be flushed.
};

clock.withMock(async () => {
  const rawFs = getFileSystem();
  const fs = new CachedFileSystem(rawFs);
  const lockFile = new LockFileUnderTest(fs);

  fs.writeFile(lockFile.lockFilePath, '188');
  const promise = lockFile.lock(async () => lockFile.log.push('fn()'));
  expect(lockFile.log).toEqual(['create()']);
  expect(lockFile.getLogger().logs.info).toEqual([...]);

  rawFs.writeFile(lockFile.lockFilePath, '444');
  await moveToNextIteration();
  await moveToNextIteration();
  expect(lockFile.getLogger().logs.info).toEqual([...]]);

  ...
});
```

[1]: https://jasmine.github.io/api/3.5/Clock

kara commented:
@petebacondarwin Are you still waiting on a review from @alxhub or is this ready to presubmit/merge?

petebacondarwin commented:
Just a presubmit please @kara
